### DISCUSSION TITLE
cuid() + uuid() in @default() for @id

### DISCUSSION THREAD

**Discussion by naeem-qv** [prisma/prisma#23307] [Q&A] [answered]
### Question

is there anyway we can combine multiple uuid() and cuid() to generate a long and unique IDs??

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id        String     @id @default(uuid())
  email     String     @unique
  name      String
  password  String
  createdAt DateTime   @default(now())
  updatedAt DateTime   @updatedAt
}
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: windows 10
- Database: PostgreSQL
- Node.js version: 20.9.0

- Run `prisma -v` to see your Prisma version and paste it
```
prisma                  : 5.10.2
@prisma/client          : 5.10.2
Computed binaryTarget   : windows
Operating System        : win32
Architecture            : x64
Node.js                 : v20.9.0
Query Engine (Node-API) : libquery-engine 5a9203d0590c951969e85a7d07215503f4672eb9 (at node_modules\.pnpm\@prisma+engines@5.10.2\node_modules\@prisma\engines\query_engine-windows.dll.node)
Schema Engine           : schema-engine-cli 5a9203d0590c951969e85a7d07215503f4672eb9 (at node_modules\.pnpm\@prisma+engines@5.10.2\node_modules\@prisma\engines\schema-engine-windows.exe)
Schema Wasm             : @prisma/prisma-schema-wasm 5.10.0-34.5a9203d0590c951969e85a7d07215503f4672eb9
Default Engines Hash    : 5a9203d0590c951969e85a7d07215503f4672eb9
Studio                  : 0.499.0
```

**Comment by ludralph** [comment, marked as answer]
Hi @naeem-qv 👋 

Thank you for raising this question.

Prisma's schema syntax allows for the use of either `uuid()` or `cuid()` as functions to generate unique identifiers for model fields marked with `@id`, but there's no documented feature or syntax for combining these functions directly in the schema definition.

However, if you're looking to create a custom `ID` that combines elements of both `uuid` and `cuid`, or any other custom logic, you might consider implementing this logic in your application code. For instance, you could generate a `uuid` and a `cuid` separately and then concatenate them in your application before inserting the record into the database. This approach would require handling the `ID` generation in your application logic.

for example:
```js
import { PrismaClient } from '@prisma/client';
import { v4 as uuidv4 } from 'uuid';
import cuid from 'cuid';

const prisma = new PrismaClient();

async function createUser(email, name, password) {
  const customId = uuidv4() + '_' + cuid();
  const user = await prisma.user.create({
    data: {
      id: customId,
      email,
      name,
      password,
      // Other fields as necessary
    },
  });
  return user;
}
```

In this example, `uuidv4()` generates a `UUID` and `cuid()` generates a `CUID`, which are then concatenated with an underscore between them to form a custom ID. This custom ID is then used when creating a new user record. If you decide to generate `IDs` in this manner, you should ensure that your Prisma model's `id` field is configured to accept strings of sufficient length to accommodate the combined ID value. 

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by naeem-qv** [reply]
Hi @ludralph ,
Thank you for your response, I will create my custom IDs in the way shown in the answer. I was just curious about knowing if Prisma has some built-in support for things like this or not. once again thanks for clearing up my doubts.
